File: 912.cpp
struct Solution {
    static constexpr int offset = 50000;
    vector<int> sortArray(vector<int>& nums) {
        for (int msk = 0; msk < 5; ++msk){
            vector<int> buckets[16];
            for (int x : nums){
                buckets[((x + this->offset) >> (msk << 2)) & 15].push_back(x);
            }
            int w = 0;
            for (int b = 0; b < 16; ++b){
                for (int x : buckets[b]){
                    nums[w++] = x;
                }
            }
        }
        return nums;
    }
};

File: sol/912.py
class Solution:
    offset = 50000
    def sortArray(self, nums: List[int]) -> List[int]:
        msk = 0
        for _ in range(5):
            buckets = [list() for _ in range(16)]
            for x in nums:
                buckets[((x + self.offset) >> (msk << 2)) & 15].append(x)
            nums = [x for b in buckets for x in b]
            msk += 1
        return nums
